### PR TITLE
Correctly handle spurious wake-ups in Map::pause()

### LINKED ISSUE #2522: Spurious wakeups
We're using `std::condition_variable` to ensure `MapContext` pauses. However, [`wait()`](http://en.cppreference.com/w/cpp/thread/condition_variable/wait) sometimes spuriously wakes up without having received an actual notification. In other words, there is no 1:1 mapping between a `notify` call and a `wait` being canceled. We should add a boolean state variable as shown in the example to make sure that we don't unpause too early.

/cc @ljbade @bleege

### PR DESCRIPTION
**Pull request by ljbade**
Fixes #2522

### CODE REVIEW

**Comment by ljbade** [comment]
This works on my phone.

File: src/mbgl/map/map.cpp [hunk 1]
```diff
@@ -32,21 +32,20 @@ Map::~Map() {
 void Map::pause() {
     assert(data->mode == MapMode::Continuous);
 
-    if (!paused) {
+    if (!data->paused) {
```
Comment (kkaefer) [inline src/mbgl/map/map.cpp]: access to `paused` needs to be within the lock.

File: src/mbgl/map/map.cpp [hunk 2]
```diff
@@ -32,21 +32,20 @@ Map::~Map() {
...
 
 bool Map::isPaused() {
-    return paused;
+    return data->paused;
 }
 
 void Map::resume() {
-    data->condResume.notify_all();
-    paused = false;
+    data->paused = false;
```
Comment (kkaefer) [inline src/mbgl/map/map.cpp]: This needs to be protected by a lock.

**Comment by ljbade** [comment]
@kkaefer I will fix the missing locks.

**Pull request merged**